enhancements/machine-api/remediation-history.md

```diff
@@ -0,0 +1,217 @@
+approvers:
+  - "@JoelSpeed"
+  - "@michaelgugino"
+  - "@enxebre"
+creation-date: 2020-12-15
+last-updated: 2021-02-08
+status: implementable
+---
+
+# Remediation history
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Record and show limited remediation history.
+
+## Motivation
+
+With machine health checks administrators can configure automatic remediation of node problems. There is no good
+overview though for when and why nodes were remediated. Recording and showing a history of remediations would allow
+administrators to review when and why machines were remediated, and to discover patterns like e.g. machineA always
+reboots at 2am on Wednesdays.
+
+### Goals
+
+- Record a limited remediation history.
+- Show remediation history in the UI.
+
+### Non-Goals
+
+- Record an *unlimited* history of remediations.
+- Collect or provide information on *why* a node has a certain node condition which triggered remediation.
+
+## Proposal
+
+The machine health check holds the configuration for automatic remediation of nodes. We want to extend the status of
+machine health checks to contain the last x remediation events, with information about which node is affected, which
+condition triggered the remediation, and the timestamps of
+- when the triggering condition was detected
+- when remediation was started
+- when the node was successfully fenced (deleted / powered off)
+
+This information can be displayed in the UI in a table on a machine healthcheck details page.
+
+MHC type enhancement:
+
+```go
+
+// MachineHealthCheckStatus defines the observed state of MachineHealthCheck
+type MachineHealthCheckStatus struct {
+
+	[...]
+
+	// History of remediations triggered by this machine health check
+	RemediationHistory []Remediation `json:"remediationHistory,omitempty"`
+}
+
+// Remediation tracks a remediation triggered by this machine health check
+type Remediation struct {
+	// a reference to the target machine or node
+	// +kubebuilder:validation:Type=string
+	Target *corev1.ObjectReference `json:"target"`
+
+	// the condition type which triggered this remediation
+	// +kubebuilder:validation:Type=string
+	ConditionType *corev1.NodeConditionType `json:"conditionType,omitempty"`
+
+	// the condition status which triggered this remediation
+	// +kubebuilder:validation:Type=string
+	ConditionStatus *corev1.ConditionStatus `json:"conditionStatus,omitempty"`
+
+	// the reason for the remediation if not a condition
+	// +kubebuilder:validation:Type=string
+	Reason string `json:"reason,omitempty"`
+
+	// the time when the unhealthy condition was detected
+	Detected *metav1.Time `json:"detected"`
```

Comment:
What happens if it goes back to healthy before the timeout? Is there some way we signal these near misses?

Reply:
I'd assume the remediation CR wouldn't be created in this case, that only happens once MHC initiates recovery (at which point it can backfill the Detected value based on the last change time of the condition)

Reply:
Good question. I like the idea of a having a `Type` for this. Maybe `aborted` instead of recovered, that would be a better fit for starting nodes than `recovered`?

Reply:
> the remediation CR wouldn't be created in this case

yes, but for built in remediation we never have a remediation CR, so I'm not sure how this answers the question?

Reply:
`aborted` works for me